add ontfs native contract

qingche123 opened the pull request:
* Add ontfs native contract

* Add multi files error return support.

* Add snarkPdp support

* Export CheckPdpProve for ontfs.

* Minimal changes.

* Update Makefile to enable cgo.

* Fix fileInfoList Deserialization function.

* Update AddErrorsEvent function.

* Add debug info.

* Add FirstPdp param with StoreFile.

* Disable FsSetGlobalParam interface.

* Update vk to adapt 4k.

* Add Challenge&Response function.

* DeleteFile when the last fileProve.

* Add GetChallenge&Judge function.

* Update Challenge/Response process code.

* Update FsJudge function.

* Move FsResponse and FsGetChallengeList function to node_business.go

* Fix build error.

* Update FsChallenge function and fix CleanFlag bug.

* Update FsChallenge function.

* Update clean fileinfo process code

* Fix FsGetChallengeList issue about checkWitness.

* Fix FsResponse deserialization issue.

* Add FsGetFileChallengeList function.

* Added judgment with challenge when the state is Judged.

* Challenge will report error when initiator is not the file owner.

* Fix the issue that pdp info is deleted and then added when all nodes submit last pdp.

* Allow readPledge settle when file has been deleted.

* File store time must be longer than 4hours.

Signed-off-by: Guo XiaoJie <guoxiaojie@onchain.com>

laizy commented on smartcontract/service/native/ontfs/passport.go:
```diff
@@ -0,0 +1,114 @@
+	if signature.Verify(pubKey, sink.Bytes(), signValue) {
+		return passport.WalletAddr, nil
+	} else {
+		return passport.WalletAddr, fmt.Errorf("CheckPassport Verify error: %s", err.Error())
```
calling `err.Error` will cause panic here,  since err is nil.

laizy commented on smartcontract/service/native/ontfs/result_info.go:
```diff
@@ -0,0 +1,60 @@
+func DecRet(ret []byte) *RetInfo {
+	var retInfo RetInfo
+	source := common.NewZeroCopySource(ret)
+	retInfo.Deserialization(source)
```
handle error

laizy commented on smartcontract/service/native/ontfs/app_sdk.go:
```diff
@@ -0,0 +1,940 @@
+	for _, addr := range nodeList {
+		nodeInfo := getNodeInfo(native, addr)
+		if nodeInfo == nil {
+			fmt.Errorf("[APP SDK] FsGetNodeInfoList getNodeInfo(%v) error", addr)
```
error created and not used. do you mean `log.Errorf`?

laizy commented on smartcontract/service/native/ontfs/node_info.go:
```diff
@@ -0,0 +1,238 @@
+	iter := native.CacheDB.NewIterator(nodeInfoPrefix[:])
+	for has := iter.First(); has; has = iter.Next() {
+		key := iter.Key()
+		item, err := utils.GetStorageItem(native, iter.Key())
```
you can use iter.Value() directly. and call `GetValueFromRawStorageItem`

laizy commented on smartcontract/service/native/ontfs/node_govern.go:
```diff
@@ -0,0 +1,211 @@
+
+	newNodeInfo.Pledge = newNodePledge
+	newNodeInfo.Profit = oldNodeInfo.Profit
+	newNodeInfo.RestVol = oldNodeInfo.RestVol + newNodeInfo.Volume - oldNodeInfo.Volume
```
use ` newNodeInfo.Volume - (oldNodeInfo.Volume - oldNodeInfo.RestVol)` to avoid overflow?

laizy commented on smartcontract/service/native/ontfs/node_govern.go:
```diff
@@ -0,0 +1,211 @@
+	return utils.BYTE_TRUE, nil
+}
+
+func FsNodeWithDrawProfit(native *native.NativeService) ([]byte, error) {
```
change `WithDraw` to `Withdraw`?

laizy commented on smartcontract/service/native/ontfs/app_sdk.go:
```diff
@@ -0,0 +1,940 @@
+
+	nodeList := getNodeAddrList(native)
+	if nodeList != nil {
+		r := rand.New(rand.NewSource(time.Now().Unix()))
```
nondeterministic execution can easily shutdown the whole network. you can use the current block hash to generate random number.

laizy commented on smartcontract/service/native/ontfs/app_sdk.go:
```diff
@@ -0,0 +1,940 @@
+			continue
+		}
+		nodesInfoList.NodesInfo = append(nodesInfoList.NodesInfo, *nodeInfo)
+		count--
```
check `count != 0` before entering for loop to avoid underflow?

laizy commented on smartcontract/service/native/ontfs/app_sdk.go:
```diff
@@ -0,0 +1,940 @@
+		}
+		nodesInfoList.NodesInfo = append(nodesInfoList.NodesInfo, *nodeInfo)
+		count--
+		if count <= 0 {
```
`count` is uint64, maybe use `==` to make it more clear.

laizy commented on smartcontract/service/native/ontfs/file_pdp_record.go:
```diff
@@ -0,0 +1,258 @@
+	return item.Value
+}
+
+func getPdpRecordMap(native *native.NativeService, fileHash []byte, fileOwner common.Address) map[common.Address]*PdpRecord {
```
clean unused function.

laizy commented on smartcontract/service/native/ontfs/app_sdk.go:
```diff
@@ -0,0 +1,940 @@
+		if !pdpRecord.SettleFlag {
+			nodeInfo.RestVol += fileInfo.FileBlockCount * DefaultPerBlockSize
+			addNodeInfo(native, nodeInfo)
+			pdpRecord.SettleFlag = true
```
seems `pdpRecord.SettleFlag` is set but not used, need update it to db?

Commit:
every file only need proof two times.

     * Remove whitelist interface.

     * Add merkle_pdp with ontfs.

     * Every file only need proof two times.

     * Increase punishAmount with Judge and FsResponse function.

     * Remove 4 hours limit with time expired in FsStoreFiles function.

     * Add DefaultOnceReadSettleGasFee.

     * Fix serverPdpGasFee error of computation.

     * Make rate of storage fee adjustable.

     * Add MinTimeForFileStorage with globalParam.

     * Fix const 'Minute' value error issue

Signed-off-by: Guo XiaoJie <guoxiaojie@onchain.com>

Commit:
Fix some errors.

Fix some errors.

Signed-off-by: Guo XiaoJie <guoxiaojie@onchain.com>

Commit:
Fix space fee bugs.

Fix space fee bugs.

Signed-off-by: Guo XiaoJie <guoxiaojie@onchain.com>

laizy commented on smartcontract/service/native/ontfs/init.go:
```diff
@@ -0,0 +1,89 @@
+func FsSetGlobalParam(native *native.NativeService) ([]byte, error) {
+	var globalParam FsGlobalParam
+	infoSource := common.NewZeroCopySource(native.Input)
+	if err := globalParam.Deserialization(infoSource); err != nil {
```
need check witness, or clean this unused code.

Commit:
Replace rand algorithm with FsGetNodeInfoList.

Replace rand algorithm with FsGetNodeInfoList.

Signed-off-by: Guo XiaoJie <guoxiaojie@onchain.com>

laizy commented on smartcontract/service/native/ontfs/node_business.go:
```diff
@@ -0,0 +1,374 @@
+}
+
+func checkPdpData(native *native.NativeService, pdpData *PdpData, fileInfo *FileInfo) error {
+	blockHeader, err := native.Store.GetHeaderByHeight(uint32(pdpData.ChallengeHeight))
```
call `GetBlockHash` instead.

laizy commented on smartcontract/service/native/ontfs/pdp/pdp.go:
```diff
@@ -0,0 +1,169 @@
+				return fmt.Errorf("[VerifyProofWithUniqueId] error: %s", err.Error())
+			}
+
+			blockId := blocksId[chl*merkle_pdp.MerkleRootLen : (chl+1)*merkle_pdp.MerkleRootLen]
```
missing `blocksId` length check.

Commit:
Clean unused code and fix some issue.

Clean unused code and fix some issue.

Signed-off-by: Guo XiaoJie <guoxiaojie@onchain.com>

laizy commented on smartcontract/service/native/ontfs/pdp/merkle_pdp/merkle_pdp.go:
```diff
@@ -0,0 +1,117 @@
+}
+
+func VerifyMerkleProof(proof [][]byte, uniqueId []byte, i uint64) error {
+	root := proof[len(proof)-1]
```
may cause out of bound panic.

laizy commented on smartcontract/service/native/ontfs/pdp/merkle_pdp/merkle_pdp.go:
```diff
@@ -0,0 +1,117 @@
+		return nil, fmt.Errorf("data length should be multiple of 32")
+	}
+
+	res := [][]byte{data[c*32 : (c+1)*32]}
```
may cause out of bound access.

laizy commented on smartcontract/service/native/ontfs/pdp/merkle_pdp/merkle_pdp.go:
```diff
@@ -0,0 +1,117 @@
+	if len(block) < 64 {
+		copy(data, block[:])
+	} else {
+		copy(data, block[0:64])
```
why only use the first 64byte?

laizy commented on smartcontract/service/native/ontfs/pdp/pdp.go:
```diff
@@ -0,0 +1,177 @@
+			binary.LittleEndian.PutUint64(vAdded, i)
+			hash := sha256.Sum256(append(plant, vAdded...))
+			bigTmp := new(big.Int).SetBytes(hash[:])
+			challenge[i] = bigTmp.Mod(bigTmp, blockNum).Uint64()
```
remove duplicated challenge number?

qingche123 replied:
“vAdded” changes every time in the loop.  So challenge number is changing

laizy replied:
you are using `mod` operation, `a != b` , but `a % blockNum == b % blockNum` can be true.

laizy commented on smartcontract/service/native/ontfs/app_sdk.go:
```diff
@@ -0,0 +1,948 @@
+	//oriPlan ==> newPlan
+	var totalAddMaxBlockNumToRead uint64
+	for index, readPlan := range readPledge.ReadPlans {
+		totalAddMaxBlockNumToRead += readPlan.MaxReadBlockNum
```
overflow check

laizy commented on smartcontract/service/native/ontfs/app_sdk.go:
```diff
@@ -0,0 +1,948 @@
+	}
+
+	newPledgeFee := totalAddMaxBlockNumToRead * globalParam.FeePerBlockForRead
+	readPledge.RestMoney += newPledgeFee
```
make sure all operations have overflow check

Commit:
Add overflow check.

Add overflow check.

Signed-off-by: Guo XiaoJie <guoxiaojie@onchain.com>

Commit:
Update merkle pdp algorithm.

Update merkle pdp algorithm.

Signed-off-by: Guo XiaoJie <guoxiaojie@onchain.com>

Commit:
Let the storage fee be calculated on an hourly basis

Let the storage fee be calculated on an hourly basis

Signed-off-by: Guo XiaoJie <guoxiaojie@onchain.com>

laizy commented on smartcontract/service/native/ontfs/pdp/pdp.go:
```diff
@@ -0,0 +1,170 @@
+
+//VerifyProof used in consensus algorithm
+func VerifyProofWithUniqueId(uniqueId []byte, proof []byte, challenge []uint64) error {
+	proofPrefix := proof[0:VersionLength]
```
the length of `proof` is still not checked.

Commit:
Remove dumplicate challenge number and add length check with VerifyProofWithUniqueId

Remove dumplicate challenge number and add length check with VerifyProofWithUniqueId

Signed-off-by: Guo XiaoJie <guoxiaojie@onchain.com>

laizy commented on smartcontract/service/native/ontfs/pdp/pdp.go:
```diff
@@ -69,25 +69,35 @@ func (p *Pdp) GenUniqueIdWithFileBlocks(fileBlocks []types.Block) ([]byte, error
+			if !sameChallenge {
+				challenge = append(challenge, challengeTmp)
+			}
+			if uint64(len(challenge)) >= challengeNum {
```
will cause dead loop if `challengeNum > blockNum`

Commit:
add ontfs native contract

* Add ontfs native contract

* Add multi files error return support.

* Add snarkPdp support

* Export CheckPdpProve for ontfs.

* Minimal changes.

* Update Makefile to enable cgo.

* Fix fileInfoList Deserialization function.

* Update AddErrorsEvent function.

* Add debug info.

* Add FirstPdp param with StoreFile.

* Disable FsSetGlobalParam interface.

* Update vk to adapt 4k.

* Add Challenge&Response function.

* DeleteFile when the last fileProve.

* Add GetChallenge&Judge function.

* Update Challenge/Response process code.

* Update FsJudge function.

* Move FsResponse and FsGetChallengeList function to node_business.go

* Fix build error.

* Update FsChallenge function and fix CleanFlag bug.

* Update FsChallenge function.

* Update clean fileinfo process code

* Fix FsGetChallengeList issue about checkWitness.

* Fix FsResponse deserialization issue.

* Add FsGetFileChallengeList function.

* Added judgment with challenge when the state is Judged.

* Challenge will report error when initiator is not the file owner.

* Fix the issue that pdp info is deleted and then added when all nodes submit last pdp.

* Allow readPledge settle when file has been deleted.

* File store time must be longer than 4hours.

Signed-off-by: Guo XiaoJie <guoxiaojie@onchain.com>
